packages/upload/src/vaadin-upload-file-list.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * @license
+ * Copyright (c) 2016 - 2022 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import './vaadin-upload-file.js';
+import { html as legacyHtml, PolymerElement } from '@polymer/polymer/polymer-element.js';
+import { html, render } from 'lit';
+import { timeOut } from '@vaadin/component-base/src/async.js';
+import { Debouncer } from '@vaadin/component-base/src/debounce.js';
+import { ThemableMixin } from '@vaadin/vaadin-themable-mixin/vaadin-themable-mixin.js';
+
+/**
+ * An element used internally by `<vaadin-upload>`. Not intended to be used separately.
+ *
+ * @extends HTMLElement
+ * @mixes FocusMixin
+ * @private
+ */
+class UploadFileList extends ThemableMixin(PolymerElement) {
+  static get is() {
+    return 'vaadin-upload-file-list';
+  }
+
+  static get template() {
+    return legacyHtml`
+      <style>
+        :host {
+          display: block;
+        }
+
+        :host([hidden]) {
+          display: none !important;
+        }
+
+        [part='list'] {
+          padding: 0;
+          margin: 0;
+          list-style-type: none;
+        }
+      </style>
+      <ul part="list">
+        <slot></slot>
+      </ul>
+    `;
+  }
+
+  static get properties() {
+    return {
+      /**
+       * The array of files being processed, or already uploaded.
+       */
+      items: {
+        type: Array,
+      },
+
+      /**
+       * The object used to localize upload files.
+       */
+      i18n: {
+        type: Object,
+      },
+    };
+  }
+
+  static get observers() {
+    return ['__updateItems(items, i18n)'];
+  }
+
+  /** @private */
+  __updateItems(items, i18n) {
+    if (items && i18n) {
+      this.__debounceUpdate = Debouncer.debounce(this.__debounceUpdate, timeOut.after(0), () => {
```

Comment:
For some reason, without debounce removing files by clicking the icon does not work properly: 
- on first "remove" click, the `vaadin-upload-file` element is not removed from DOM,
- on "remove" click for another file, both `vaadin-upload-file` elements are removed.

Reply:
The setup for removing a file is pretty wacky:
- `vaadin-upload-file` sends a `file-abort` event
- `vadin-upload` handles that, changes the file state to aborted, then updates the file list / triggers a first render
- `vaadin-upload-file` has an observer for `aborted`, then dispatches a `file-remove` event
- `vadin-upload` handles that, then actually removes the file, triggers a second render of the file list

Maybe we could remove the file as part of handling the initial abort event? Something like this: https://github.com/vaadin/web-components/commit/e251fa935608021f669ec19dd43c1ec22b1e90ac

Reply:
> The setup for removing a file is pretty wacky

IIRC, the reason for the current behavior was to be able to animate removal of the file.
It's even mentioned in JSDoc and there is an issue to get rid of that: #2313

Let me create a separate PR for changing this logic to fire `file-remove` immediately.

Reply:
One caveat is that there is a `GeneratedVaadinUploadFile` Flow component that expects the source of a `file-remove` event to be the upload file, not the upload component itself. But it doesn't look like the component is actually used, and I can not find any other reference to `file-remove`.

Reply:
Done in #4894.